session.go

```diff
@@ -2254,42 +2254,74 @@ type traceWriter struct {
 	session *Session
 	w       io.Writer
 	mu      sync.Mutex
+
+	tracingInfoFetchAttempts int
+	tracingInfoFetchInterval time.Duration
 }
 
 // NewTraceWriter returns a simple Tracer implementation that outputs
 // the event log in a textual format.
 func NewTraceWriter(session *Session, w io.Writer) Tracer {
-	return &traceWriter{session: session, w: w}
+	return &traceWriter{session: session, w: w, tracingInfoFetchAttempts: 5, tracingInfoFetchInterval: 3 * time.Millisecond}
 }
 
 func (t *traceWriter) Trace(traceId []byte) {
+	var (
+		timestamp time.Time
+		activity  string
+		source    string
+		elapsed   int
+		thread    string
+	)
+
+	t.mu.Lock()
+	defer t.mu.Unlock()
+
+	fetchAttempts := 1
+	if t.tracingInfoFetchAttempts > 0 {
+		fetchAttempts = t.tracingInfoFetchAttempts
+	}
+
+	for i := 0; i < fetchAttempts; i++ {
+		isDone := false
+		iter := t.session.control.query(`SELECT event_id, activity, source, source_elapsed, thread
+			FROM system_traces.events
+			WHERE session_id = ?`, traceId)
+
+		for iter.Scan(&timestamp, &activity, &source, &elapsed, &thread) {
+			if strings.Contains(activity, "Done processing") {
+				isDone = true
+				break
+			}
+		}
+
+		if err := iter.Close(); err != nil {
+			fmt.Fprintln(t.w, "Error:", err)
+			return
+		}
+
+		if isDone || i == fetchAttempts-1 {
+			break
+		}
+
+		time.Sleep(t.tracingInfoFetchInterval)
+	}
```

Comment:
```suggestion
	isDone := false
	outer:
	for i := 0; i < fetchAttempts; i++ {
		iter := t.session.control.query(`SELECT event_id, activity, source, source_elapsed, thread
			FROM system_traces.events
			WHERE session_id = ?`, traceId)

		for iter.Scan(&timestamp, &activity, &source, &elapsed, &thread) {
			if strings.Contains(activity, "Done processing") {
				isDone = true
				break outer
			}
		}

		if err := iter.Close(); err != nil {
			fmt.Fprintln(t.w, "Error:", err)
			return
		}

		time.Sleep(t.tracingInfoFetchInterval)
	}
	if !isDone {
		fmt.Fprintln(t.w, "Error: failed to wait tracing to complete. !!! Tracing is incomplete !!!")
	}
```

Reply:
one problem with this approach: when we break outer the iter.Close() is not executed

Reply:
take a look again